reasoner/planner/HybridInformedPlanner.java

```diff
@@ -0,0 +1,287 @@
+        super(traversalEng, conceptMgr, logicMgr);
+    }
+
+    @Override
+    Set<OrderingCoster.LocalAllCallsCosting>  recursivelyGenerateCostings(CallMode callMode) {
+        HybridAStarBeamSearch ucs = new HybridAStarBeamSearch(this, callMode);
+        ucs.search();
+        return iterate(ucs.completedSignatures().values())
+                .map(fringeElement -> fringeElement.costingBuilder.build())
+                .map(singleCallCosting -> orderingCoster.createAllCallsCosting(singleCallCosting.callMode, singleCallCosting.ordering, singleCallCosting.cyclicConcludableModes))
+                .toSet();
+    }
+
+    private static class HybridAStarBeamSearch {
+        private final HybridInformedPlanner planner;
+        private final ReasonerPlanner.CallMode callMode;
+
+        private final ConjunctionGraph.ConjunctionNode conjunctionNode;
+        private final Set<Resolvable<?>> resolvables;
+        private final Set<Concludable> cyclicConcludables;
+        private final Map<Resolvable<?>, Set<Variable>> dependencies;
+
+        private final Map<Set<Pair<Concludable, Set<Variable>>>, FringeElement> completedSignatures;
+        private final Set<Resolvable<?>> valueIdentityResolvables;
+        private final Map<Resolvable<?>, Double> resolvableCostForHeuristic;
+
+        public HybridAStarBeamSearch(HybridInformedPlanner planner, ReasonerPlanner.CallMode callMode) {
+            this.planner = planner;
+            this.callMode = callMode;
+            this.conjunctionNode = planner.conjunctionGraph.conjunctionNode(callMode.conjunction);
+            this.resolvables = planner.logicMgr.compile(callMode.conjunction);
+            this.cyclicConcludables = conjunctionNode.cyclicConcludables();
+            this.dependencies = ReasonerPlanner.dependencies(resolvables);
+            this.valueIdentityResolvables = iterate(this.resolvables).filter(resolvable -> !resolvable.isNegated()).filter(r ->
+                    iterate(estimateableVariables(r.variables())).flatMap(v -> iterate(v.constraints()))
+                            .anyMatch(constraint ->
+                                    constraint.isThing() && constraint.asThing().isValue() && constraint.asThing().asValue().isValueIdentity())).toSet();
+
+            this.resolvableCostForHeuristic = planner.orderingCoster.resolvableHeuristics(callMode, conjunctionNode, resolvables);
+            this.completedSignatures = new HashMap<>();
+        }
+
+
+        public Map<Set<Pair<Concludable, Set<Variable>>>, FringeElement> completedSignatures() {
+            return completedSignatures;
+        }
+
+        public void search() {
+            PriorityQueue<FringeElement> fringe = new PriorityQueue<>();
+            fringe.add(initialFringeElement());
+            fringe = aStarSearch(fringe, new HashSet<>());
+            fringe = beamSearch(fringe, new HashSet<>()); // Start with a fresh seen set.
```

Comment:
I do reuse the old fringe, but this makes little sense because longer plans in the existing fringe will be favoured for no reason